// pi-camera-connect/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var still_camera_1 = require("./lib/still-camera");
exports.StillCamera = still_camera_1.default;
var stream_camera_1 = require("./lib/stream-camera");
exports.StreamCamera = stream_camera_1.default;
exports.Codec = stream_camera_1.Codec;
exports.SensorMode = stream_camera_1.SensorMode;
var Rotation;
(function (Rotation) {
    Rotation[Rotation["Rotate0"] = 0] = "Rotate0";
    Rotation[Rotation["Rotate90"] = 90] = "Rotate90";
    Rotation[Rotation["Rotate180"] = 180] = "Rotate180";
    Rotation[Rotation["Rotate270"] = 270] = "Rotate270";
})(Rotation = exports.Rotation || (exports.Rotation = {}));
var Flip;
(function (Flip) {
    Flip["None"] = "none";
    Flip["Horizontal"] = "horizontal";
    Flip["Vertical"] = "vertical";
    Flip["Both"] = "both";
})(Flip = exports.Flip || (exports.Flip = {}));
var ExposureMode;
(function (ExposureMode) {
    ExposureMode["Off"] = "off";
    ExposureMode["Auto"] = "auto";
    ExposureMode["Night"] = "night";
    ExposureMode["NightPreview"] = "nightpreview";
    ExposureMode["Backlight"] = "backlight";
    ExposureMode["Spotlight"] = "spotlight";
    ExposureMode["Sports"] = "sports";
    ExposureMode["Snow"] = "snow";
    ExposureMode["Beach"] = "beach";
    ExposureMode["VeryLong"] = "verylong";
    ExposureMode["FixedFps"] = "fixedfps";
    ExposureMode["AntiShake"] = "antishake";
    ExposureMode["Fireworks"] = "fireworks";
})(ExposureMode = exports.ExposureMode || (exports.ExposureMode = {}));
var AwbMode;
(function (AwbMode) {
    AwbMode["Off"] = "off";
    AwbMode["Auto"] = "auto";
    AwbMode["Sun"] = "sun";
    AwbMode["Cloud"] = "cloud";
    AwbMode["Shade"] = "shade";
    AwbMode["Tungsten"] = "tungsten";
    AwbMode["Fluorescent"] = "fluorescent";
    AwbMode["Incandescent"] = "incandescent";
    AwbMode["Flash"] = "flash";
    AwbMode["Horizon"] = "horizon";
    AwbMode["GreyWorld"] = "greyworld";
})(AwbMode = exports.AwbMode || (exports.AwbMode = {}));
